### PR TITLE
Single webpack entry

### PR DESCRIPTION
**Pull request by sndrs**
## What does this change?

- creates a single[^1] webpack entry for the JS application
- enables webpack's `splitChunks` optimisation to remove inefficient duplication across chunks
- creates a new 'frameworks' chunk as of `splitChunks` that contains the larger 3rd party libs we know we cannot run without (e..g preact, emotion) and others that need to be singletons (e.g. swr) 

[^1]: _ga needs a slightly more complicated refactor #7093_

## Why?

_tl;dr_ 
it makes it much easier to reason about how and when we use code, and knocks 93kB (gzipped) of the build.

| Before      | After      |
|-------------|------------|
| 579.15 KB | 486.37kB |

The current strategy, introduced in #727, means that webpack cannot make the most of its code-splitting/tree-shaking algorithms, because it cannot trace the import map of the entire application. 

With only one entry per page, webpack can chunk subsequent imports in a way that takes most advantage of available bandwidth, reuse of previous imports and so on.

From [the docs](https://webpack.js.org/guides/code-splitting/#prevent-duplication):

> Although using multiple entry points per page is allowed in webpack, it should be avoided when possible in favor of an entry point with multiple imports: entry: { page: ['./analytics', './app'] }. This results in a better optimization and consistent execution order when using async script tags.

### CODE REVIEW

File: dotcom-rendering/src/web/server/articleToHtml.tsx [hunk 1]
```diff
@@ -125,23 +122,6 @@ export const articleToHtml = ({ article }: Props): string => {
...
-	const lowPriorityScriptTags = generateScriptTags(
-		[
-			...getScriptArrayFromFile('atomIframe.js'),
-			...getScriptArrayFromFile('embedIframe.js'),
-			...getScriptArrayFromFile('newsletterEmbedIframe.js'),
-			...getScriptArrayFromFile('relativeTime.js'),
-			...getScriptArrayFromFile('initDiscussion.js'),
-		].map((script) => (offerHttp3 ? getHttp3Url(script) : script)),
-	);
-
```
Comment (mxdvl) [inline dotcom-rendering/src/web/server/articleToHtml.tsx:144]: Begone!

File: dotcom-rendering/scripts/webpack/webpack.config.browser.js [hunk 2]
```diff
@@ -98,19 +98,33 @@ const getLoaders = (bundle) => {
...
+				defaultVendors: {
+					test: /[\\/]node_modules[\\/]/,
+					priority: -10,
+					reuseExistingChunk: true,
+				},
+				default: {
+					minChunks: 2,
+					priority: -20,
+					reuseExistingChunk: true,
+				},
```
Comment (mxdvl) [inline dotcom-rendering/scripts/webpack/webpack.config.browser.js:125]: Should we add a bit more information about where to find documentation on this stuff? Feels like a lot of magic numbers.

File: dotcom-rendering/src/web/browser/startup.ts [hunk 3]
```diff
@@ -12,9 +13,11 @@ const measure = (name: string, task: () => Promise<void>): void => {
 		// See: https://github.com/cypress-io/cypress/issues/2651#issuecomment-432698837
 		.then(() => {
 			end();
+			log('dotcom', `🥾 Booted ${name} in ${end()}ms`);
```
[the remark this answers is not in the document]
Comment (sndrs) [reply dotcom-rendering/src/web/browser/startup.ts:16]: 💯

File: dotcom-rendering/scripts/webpack/webpack.config.browser.js [hunk 4]
```diff
@@ -98,19 +98,33 @@ const getLoaders = (bundle) => {
...
+		splitChunks: {
+			cacheGroups: {
+				// our own chunk, which is shared between all bundles
+				frameworks: {
+					test: /[\\/]node_modules[\\/](preact|react-is|hoist-non-react-statistics|swr|@emotion|stylis)[\\/]/,
+					chunks: 'all',
+					name: 'frameworks',
+					enforce: true,
+				},
+				// webpack defaults
```
Comment (sndrs) [inline dotcom-rendering/scripts/webpack/webpack.config.browser.js:115]:
```suggestion
				// webpack defaults 
				// https://webpack.js.org/plugins/split-chunks-plugin/#optimizationsplitchunks
```

**Comment by sndrs** [comment]
@arelra ok - can you let me know when that's done? how long do you expect it to take?

**Comment by arelra** [comment]
> @arelra ok - can you let me know when that's done? how long do you expect it to take?

1) SWC is all good: https://github.com/guardian/dotcom-rendering/pull/7110
2) I haven't heard back from D&I yet but I expect/hope that we can conclude the test by EOW

**Comment by arelra** [comment]
> @arelra ok - can you let me know when that's done? how long do you expect it to take?

Hey @sndrs the test has completed so feel free to merge now. Thanks.

**Review by OllysCoding — COMMENTED** [review]
I realise I'm a bit late to the game on this PR but it looks great :) All round the changes feel like a simplification which will help (hopefully) both get us better performance for our users, as well as make life much simpler for future developers. 

A potential future improvement for us could be to investigate the mechanisms around the `startup` script/function. The function works with a mix of code in the webpack build and code written directly in the html template, there is inconsistent behaviour and it's not particularly clearly documented!

**Comment by sndrs** [comment]
does it help to address this now?

https://github.com/guardian/dotcom-rendering/blob/aaa3737538849bc099a7274ad5fa7e5d151dcc1d/dotcom-rendering/window.guardian.d.ts#L24-L27

**Pull request merged**